pkg/receive/multitsdb.go

```diff
@@ -0,0 +1,269 @@
+	if l == nil {
+		l = log.NewNopLogger()
+	}
+
+	return &MultiTSDB{
+		dataDir:         dataDir,
+		logger:          l,
+		reg:             reg,
+		tsdbCfg:         tsdbCfg,
+		mtx:             &sync.RWMutex{},
+		dbs:             map[string]*FlushableStorage{},
+		stores:          map[string]*store.TSDBStore{},
+		appendables:     map[string]*tsdb.ReadyStorage{},
+		shippers:        map[string]*shipper.Shipper{},
+		labels:          labels,
+		tenantLabelName: tenantLabelName,
+		bucket:          bucket,
+		upload:          bucket != nil,
+	}
+}
+
+func (t *MultiTSDB) Open() error {
+	if err := os.MkdirAll(t.dataDir, 0777); err != nil {
+		return err
+	}
+
+	return t.openTSDBs()
+}
+
+func (t *MultiTSDB) Close() error {
+	t.mtx.Lock()
+	defer t.mtx.Unlock()
+
+	errmtx := &sync.Mutex{}
+	merr := terrors.MultiError{}
+	wg := &sync.WaitGroup{}
+	for _, tsdb := range t.dbs {
+		tsdb := tsdb
+		wg.Add(1)
+		go func() {
+			if err := tsdb.Close(); err != nil {
+				errmtx.Lock()
+				merr.Add(err)
+				errmtx.Unlock()
+			}
+			wg.Done()
+		}()
+	}
+
+	wg.Wait()
+	return merr.Err()
+}
+
+func (t *MultiTSDB) Flush() error {
+	t.mtx.Lock()
+	defer t.mtx.Unlock()
+
+	errmtx := &sync.Mutex{}
+	merr := terrors.MultiError{}
+	wg := &sync.WaitGroup{}
+	for _, tsdb := range t.dbs {
+		tsdb := tsdb
+		wg.Add(1)
+		go func() {
+			if err := tsdb.Flush(); err != nil {
+				errmtx.Lock()
+				merr.Add(err)
+				errmtx.Unlock()
+			}
+			wg.Done()
+		}()
+	}
+
+	wg.Wait()
+	return merr.Err()
+}
+
+func (t *MultiTSDB) Upload(ctx context.Context) error {
+	if !t.upload {
+		return nil
+	}
+
+	t.mtx.Lock()
+	defer t.mtx.Unlock()
+
+	errmtx := &sync.Mutex{}
+	merr := terrors.MultiError{}
+	wg := &sync.WaitGroup{}
+	for tenant, s := range t.shippers {
+		level.Debug(t.logger).Log("msg", "uploading block for tenant", "tenant", tenant)
+		s := s
+		wg.Add(1)
+		go func() {
+			if uploaded, err := s.Sync(ctx); err != nil {
+				errmtx.Lock()
+				merr.Add(fmt.Errorf("failed to upload %d: %w", uploaded, err))
+				errmtx.Unlock()
+			}
+			wg.Done()
+		}()
+	}
+
+	wg.Wait()
+	return merr.Err()
+}
+
+func (t *MultiTSDB) openTSDBs() error {
+	files, err := ioutil.ReadDir(t.dataDir)
+	if err != nil {
+		return err
+	}
+
+	var g errgroup.Group
+	for _, f := range files {
+		f := f
+		if !f.IsDir() {
+			continue
+		}
+
+		g.Go(func() error {
+			tenantId := f.Name()
+			_, err := t.getOrLoadTenant(tenantId)
+			return err
+		})
+	}
+
+	return g.Wait()
+}
+
+func (t *MultiTSDB) TSDBStores() map[string]*store.TSDBStore {
+	t.mtx.RLock()
+	res := make(map[string]*store.TSDBStore, len(t.stores))
+	for k, v := range t.stores {
+		res[k] = v
+	}
+	defer t.mtx.RUnlock()
+	return res
+}
+
+func (t *MultiTSDB) getOrLoadTenant(tenantID string) (*tsdb.ReadyStorage, error) {
+	// Fast path, as creating tenants is a very rare operation.
```

Comment:
Fast if rare? :thinking:  that sounds weird. So who we care about fast then?

Reply:
To me the original comment makes more sense, but no strong opinion.

Reply:
fair, now when I understood code it makes sense. But just out of nowhere for first time reader it can surprise (: Mabe just my English is broken